p2p/peer.py

```diff
@@ -598,8 +598,20 @@ def __len__(self):
     def is_full(self) -> bool:
         return len(self) >= self.max_peers
 
+    def is_valid_connection_candidate(self, candidate: Node) -> bool:
+        max_matching_ip = 2
+        matching_ip = 0
+        for peer in self.connected_nodes.values():
```

Comment:
We tend to aim for near zero mutability in our code which generally means avoiding things like `+=` or really writing to a variable more than once.  Also, avoiding things like appending to lists and mutating dictionaries.  Here's some untested code that should accomplish what you're going for here.

```python
import operator
from cytoolz import groupby

def is_valid_connection_candidate(self, candidate: Node) -> bool:
    if candidate == self.this_node:
        return False

    nodes_by_ip = groupby(operator.attrgetter('address.ip'), self.connected_nodes.values())
    matching_ip_nodes = nodes_by_ip.get(candidate.address.ip, [])
    
    return len(matching_ip_nodes) < 2
```